src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -59,36 +79,145 @@ private void combineShardScores(final ScoreCombinationTechnique scoreCombination
+        final CompoundTopDocs compoundTopDocs,
+        final boolean isSortByScore,
+        Map<Integer, Float> combinedNormalizedScoresByDocId
+    ) {
+        // we're merging docs with normalized and combined scores. we need to have only maxHits results
+        Map<Integer, Object[]> docIdSortFieldMap = new HashMap<>();
+        List<TopDocs> topFieldDocs = compoundTopDocs.getTopDocs();
+
+        for (TopDocs topDocs : topFieldDocs) {
+            for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                FieldDoc fieldDoc = (FieldDoc) scoreDoc;
+
+                if (docIdSortFieldMap.get(fieldDoc.doc) == null) {
+                    // If sort by score then replace sort field value with normalized score.
+                    if (isSortByScore) {
+                        docIdSortFieldMap.put(fieldDoc.doc, new Object[] { combinedNormalizedScoresByDocId.get(fieldDoc.doc) });
+                    } else {
+                        docIdSortFieldMap.put(fieldDoc.doc, fieldDoc.fields);
+                    }
+                }
+            }
+        }
+        return docIdSortFieldMap;
     }
 
-    private List<Integer> getSortedDocIds(final Map<Integer, Float> combinedNormalizedScoresByDocId) {
+    private List<Integer> getSortedDocIds(
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final List<TopFieldDocs> topFieldDocs,
+        final Sort sort
+    ) {
         // we're merging docs with normalized and combined scores. we need to have only maxHits results
-        List<Integer> sortedDocsIds = new ArrayList<>(combinedNormalizedScoresByDocId.keySet());
-        sortedDocsIds.sort((a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a)));
+        List<Integer> sortedDocsIds;
+        if (Objects.isNull(sort)) {
+            sortedDocsIds = new ArrayList<>(combinedNormalizedScoresByDocId.keySet());
+            sortedDocsIds.sort((a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a)));
+        } else {
+            if (Objects.isNull(topFieldDocs)) {
+                throw new IllegalArgumentException("topFieldDocs cannot be null when sorting is enabled.");
+            }
+            int topN = 0;
+            for (TopFieldDocs topFieldDoc : topFieldDocs) {
+                topN += topFieldDoc.scoreDocs.length;
+            }
+
+            // Merge the sorted results of individual queries to form a one final result per shard which is sorted.
+            final TopDocs sortedTopDocs = TopDocs.merge(sort, 0, topN, topFieldDocs.toArray(new TopFieldDocs[0]), getTieBreaker());
+
+            // Remove duplicates from the sorted top docs.
+            Set<Integer> uniqueDocIds = new LinkedHashSet<>();
+            for (ScoreDoc scoreDoc : sortedTopDocs.scoreDocs) {
+                uniqueDocIds.add(scoreDoc.doc);
+            }
+            sortedDocsIds = new ArrayList<>(uniqueDocIds);
```

Comment:
At the first glance it looks like lot of tossing of same info between different collections: List of topDocs -> array -> sorted topDocs -> set of doc ids -> list of unique ids
it may be a hot spot as we're on the search path. Can we eliminate some iterations? Why we do have duplicates in score docs?

Reply:
1. List of TopFieldDocs is input of the method.
2. Method **TopDocs.merge**
   a. Input: Array of TopFieldDocs
   b. Output: Get list of sorted score docs which contains duplicates basically results from multiple queries.
3. Convert sortedTopDocs.scoreDocs to list and then pass it to set. 

As the list is sorted so all the top entries will be added first in the set.